File: src/types/metrics.ts
export interface CpuMetrics {
  overall: number;
  cores: number[];
}

export interface RamMetrics {
  total_mb: number;
  used_mb: number;
  available_mb: number;
}

export interface DiskEntry {
  name: string;
  total_gb: number;
  used_gb: number;
}

export interface DiskMetrics {
  disks: DiskEntry[];
  total_gb: number;
  used_gb: number;
}

export interface NetworkInterface {
  name: string;
  download_kbps: number;
  upload_kbps: number;
}

export interface NetworkMetrics {
  download_kbps: number;
  upload_kbps: number;
  interfaces: NetworkInterface[];
}

export interface SystemMetrics {
  uptime_secs: number;
  swap_total_mb: number;
  swap_used_mb: number;
}

export interface TemperatureSensor {
  label: string;
  temp_c: number;
  max_c: number | null;
  critical_c: number | null;
}

export interface HardwareMetrics {
  sensors: TemperatureSensor[];
  max_temp_c: number | null;
}

export interface BatteryMetrics {
  present: boolean;
  percent: number | null;
  state: string | null;
  time_to_full_secs: number | null;
  time_to_empty_secs: number | null;
}

export interface AllMetrics {
  cpu: CpuMetrics;
  ram: RamMetrics;
  disk: DiskMetrics;
  network: NetworkMetrics;
  system: SystemMetrics;
  hardware: HardwareMetrics;
  battery: BatteryMetrics;
}

export interface MetricsHistory {
  cpu: number[];
  ram: number[];
  disk: number[];
  download: number[];
  upload: number[];
  temperature: number[];
}

export const HISTORY_LENGTH = 60;

export type SpeedUnit = "kbps" | "mbps";

export interface AppSettings {
  pollIntervalMs: number;
  speedUnit: SpeedUnit;
  alertsEnabled: boolean;
  cpuAlertThreshold: number;
  ramAlertThreshold: number;
  diskAlertThreshold: number;
  alwaysOnTop: boolean;
  autostart: boolean;
  transparentWindow: boolean;
}

export const DEFAULT_SETTINGS: AppSettings = {
  pollIntervalMs: 1000,
  speedUnit: "kbps",
  alertsEnabled: true,
  cpuAlertThreshold: 90,
  ramAlertThreshold: 90,
  diskAlertThreshold: 90,
  alwaysOnTop: false,
  autostart: false,
  transparentWindow: false,
};

File: src-tauri/src/metrics.rs
use serde::Serialize;

#[derive(Serialize)]
pub struct CpuMetrics {
    pub overall: f32,
    pub cores: Vec<f32>,
}

#[derive(Serialize)]
pub struct RamMetrics {
    pub total_mb: u64,
    pub used_mb: u64,
    pub available_mb: u64,
}

#[derive(Serialize)]
pub struct DiskEntry {
    pub name: String,
    pub total_gb: u64,
    pub used_gb: u64,
}

#[derive(Serialize)]
pub struct DiskMetrics {
    pub disks: Vec<DiskEntry>,
    pub total_gb: u64,
    pub used_gb: u64,
}

#[derive(Serialize)]
pub struct NetworkInterface {
    pub name: String,
    pub download_kbps: u64,
    pub upload_kbps: u64,
}

#[derive(Serialize)]
pub struct NetworkMetrics {
    pub download_kbps: u64,
    pub upload_kbps: u64,
    pub interfaces: Vec<NetworkInterface>,
}

#[derive(Serialize)]
pub struct SystemMetrics {
    pub uptime_secs: u64,
    pub swap_total_mb: u64,
    pub swap_used_mb: u64,
}

#[derive(Serialize)]
pub struct TemperatureSensor {
    pub label: String,
    pub temp_c: f32,
    pub max_c: Option<f32>,
    pub critical_c: Option<f32>,
}

#[derive(Serialize)]
pub struct HardwareMetrics {
    pub sensors: Vec<TemperatureSensor>,
    pub max_temp_c: Option<f32>,
}

#[derive(Serialize)]
pub struct BatteryMetrics {
    pub present: bool,
    pub percent: Option<f32>,
    pub state: Option<String>,
    pub time_to_full_secs: Option<u64>,
    pub time_to_empty_secs: Option<u64>,
}

#[derive(Serialize)]
pub struct AllMetrics {
    pub cpu: CpuMetrics,
    pub ram: RamMetrics,
    pub disk: DiskMetrics,
    pub network: NetworkMetrics,
    pub system: SystemMetrics,
    pub hardware: HardwareMetrics,
    pub battery: BatteryMetrics,
}
